X86/X86MachineInstructionRaiser.cpp

```
@@ -2447,6 +2447,140 @@ bool X86MachineInstructionRaiser::raiseDivideInstr(const MachineInstr &MI,
+  Value *Op2Val = nullptr;
+  if (Op2.isReg()) {
+    Op2Val = raisedValues->castValue(getRegOrArgValue(Op2.getReg(), MBBNo),
+                                     OpValues[0]->getType(), RaisedBB);
+  } else if (Op2.isImm()) {
+    Op2Val = ConstantInt::get(OpValues[0]->getType(), Op2.getImm());
+  } else {
+    LLVM_DEBUG(MI.dump());
+    assert(false && "Unhandled operand type in compare instruction");
+  }
+  OpValues[1] = Op2Val;
+
+  assert(OpValues[0] != nullptr && OpValues[1] != nullptr &&
+         "Unable to materialize compare operand values");
+
+  assert((OpValues[0]->getType() == OpValues[1]->getType()) &&
+         "Mis-matched operand types encountered while raising bt instruction");
+
+  auto BitWidth = OpValues[0]->getType()->getPrimitiveSizeInBits();
+  auto ModuloOperand = ConstantInt::get(OpValues[1]->getType(), BitWidth);
+  auto PositionModulo =
+      BinaryOperator::CreateURem(OpValues[1], ModuloOperand, "", RaisedBB);
+
+  auto ConstOne = ConstantInt::get(OpValues[0]->getType(), 1);
+  auto BitMask =
+      BinaryOperator::CreateShl(ConstOne, PositionModulo, "", RaisedBB);
+
+  auto BitValue = BinaryOperator::CreateAnd(OpValues[0], BitMask, "", RaisedBB);
+  auto ConstZero = ConstantInt::get(OpValues[0]->getType(), 0);
+  auto CFValue = ICmpInst::Create(Instruction::ICmp, ICmpInst::ICMP_NE,
+                                  BitValue, ConstZero, "", RaisedBB);
+
+  raisedValues->setEflagValue(EFLAGS::CF, MI.getParent()->getNumber(), CFValue);
+
+  Value *WriteBackVal = nullptr;
+  if (isBTSInstr) {
+    WriteBackVal = BinaryOperator::CreateOr(OpValues[0], BitMask, "", RaisedBB);
+  } else if (isBTRInstr) {
+    BitMask = BinaryOperator::CreateNot(BitMask, "", RaisedBB);
+    WriteBackVal =
+        BinaryOperator::CreateAnd(OpValues[0], BitMask, "", RaisedBB);
+  } else if (isBTCInstr) {
+    WriteBackVal =
+        BinaryOperator::CreateXor(OpValues[0], BitMask, "", RaisedBB);
+  }
+
+  if (WriteBackVal != nullptr) {
```

assert that `WriteBackVal != nullptr` instead of testing for it.

This will eliminate silent failure in case `WriteBackVal == nullptr`

---

For `BT` `WriteBackValue` is not set, as we don't modify the register/memory location.
Should I write the assert in the form `assert((isBTSInstr || isBTRInstr || isBTCInstr) && WriteBackVal != nullptr)`?
Or would it be better to have an early return for `BT` and just `assert(WriteBackVal != nullptr)`?

---

Good point. Ignore the suggestion to change it to assert.